spec/controllers/saml_idp_controller_spec.rb

```diff
@@ -44,6 +44,76 @@
 
       delete :logout, params: { SAMLRequest: 'foo' }
     end
+
+    let(:service_provider) do
+      create(:service_provider,
+             cert: nil, # override singular cert
+             certs: ['saml_test_sp'],
+             active: true)
+    end
+
+    let(:right_cert_settings) do
+      sp1_saml_settings.tap do |settings|
+        settings.issuer = service_provider.issuer
+      end
+    end
+
+    let(:wrong_cert_settings) do
+      sp1_saml_settings.tap do |settings|
+        settings.issuer = service_provider.issuer
+        settings.certificate = File.read(Rails.root.join('certs', 'sp', 'saml_test_sp2.crt'))
+        settings.private_key = OpenSSL::PKey::RSA.new(
+          File.read(Rails.root + 'keys/saml_test_sp2.key'),
+        ).to_pem
+      end
+    end
+
+    it 'accepts requests from a correct cert' do
+      saml_request = UriService.params(
+        OneLogin::RubySaml::Logoutrequest.new.create(right_cert_settings),
+      )[:SAMLRequest]
+
+      payload = {
+        SAMLRequest: saml_request,
+        RelayState: 'aaa',
+        SigAlg: 'SHA256',
+      }
+      canon_string = payload.to_query
+
+      private_sp_key = OpenSSL::PKey::RSA.new(right_cert_settings.private_key)
+      signature = Base64.encode64(private_sp_key.sign(OpenSSL::Digest::SHA256.new, canon_string))
+
+      certificate = OpenSSL::X509::Certificate.new(right_cert_settings.certificate)
+
+      puts "OUTSIDE GEM CANON"
+      puts canon_string
+      puts "----"
+      puts signature
+      puts "----"
+      puts certificate.serial
+      puts "----"
+
+      # This is the same verification process we expect the SAML gem will run
+      expect(
+        certificate.public_key.verify(
+          OpenSSL::Digest::SHA256.new,
+          Base64.decode64(signature),
+          canon_string,
+        ),
+      ).to eq(true)
```

Comment:
this passes, but the equivalent `.verify` inside the gem is failing: 

https://github.com/18F/saml_idp/blob/eba368170c24f2a241537565bb1be32532ac754a/lib/saml_idp/xml_security.rb#L190

all the `puts` statements confirmed of for me that the string, the signature, and the serial are all the same, and the outside verificatin works but the gem says no

Reply:
Turns out `Hash#to_query` does not preserve order, fixed in a31640f